### ISSUE TITLE
Using underscored: true still returns attributes in camelCase

### ISSUE THREAD

**Issue by serge1peshcoff** [sequelize/sequelize#10857]
## What are you doing?
I have a project that uses Sequelize as ORM. I need the attribute names returned in `snake_case`, but although I've specified `underscored: true` in my model definition, the attributes are returned in `camelCase` instead. (The DB tables are in `snake_case`, as expected).

I'm also using Sequelize 4.x in a few other projects and it doesn't affect it afaik.

Can be the same reason as here: https://github.com/sequelize/sequelize/issues/10581 (but it was closed because of not proper issue format).

I can provide more details if needed.

Sequelize config:
```js
const Sequelize = require('sequelize');

const getSequelize = () => new Sequelize(config.postgres.database, config.postgres.username, config.postgres.password, {
    host: config.postgres.host,
    port: config.postgres.port,
    dialect: 'postgres',
    logging: sql => logger.debug(sql)
});

let sequelize = getSequelize();

exports.sequelize = sequelize;
exports.Sequelize = Sequelize;
```

Model definition
```js
const { Sequelize, sequelize } = require('../lib/sequelize');

const Code = sequelize.define('code', {
    value: {
        type: Sequelize.STRING,
        allowNull: false,
        defaultValue: '',
        validate: {
            notEmpty: { msg: 'Code should be set.' },
        },
    },
    integration_id: {
        type: Sequelize.INTEGER,
        allowNull: false,
        defaultValue: '',
        validate: {
            notEmpty: { msg: 'Integration ID should be set.' },
        },
    },
    claimed_by: {
        type: Sequelize.INTEGER
    }
}, { underscored: true, tableName: 'codes' });

module.exports = Code;
```

Endpoint using this model (I have 2 models but I provide only one for the case of simplicity):
```js
exports.getMyCodes = async (req, res) => {
    const myCodes = await Code.findAll({
        where: {
            claimed_by: req.user.id
        },
        include: [Integration],
        order: [
            ['updated_at', 'DESC'],
        ]
    });

    return res.json(myCodes);
};

```

## What do you expect to happen?
Everything returned in `snake_case`

## What is actually happening?
The server response (note the `integrationId`, `createdAt` and `updatedAt` are in `camelCase`):
```js
[{
    "id": 7,
    "value": "2",
    "integration_id": 2,
    "claimed_by": 1,
    "createdAt": "2019-04-28T19:14:51.641Z",
    "updatedAt": "2019-04-28T19:16:24.259Z",
    "integrationId": 2,
    "integration": {
        "id": 2,
        "name": "Help",
        "code": "flixbus",
        "quota_period": "month",
        "quota_amount": 3,
        "description": "**I need somebody**",
        "createdAt": "2019-04-28T19:14:15.164Z",
        "updatedAt": "2019-04-28T19:14:15.164Z"
    }
}, {
    "id": 6,
    "value": "1",
    "integration_id": 2,
    "claimed_by": 1,
    "createdAt": "2019-04-28T19:14:51.641Z",
    "updatedAt": "2019-04-28T19:14:55.130Z",
    "integrationId": 2,
    "integration": {
        "id": 2,
        "name": "Help",
        "code": "flixbus",
        "quota_period": "month",
        "quota_amount": 3,
        "description": "**I need somebody**",
        "createdAt": "2019-04-28T19:14:15.164Z",
        "updatedAt": "2019-04-28T19:14:15.164Z"
    }
}, {
    "id": 1,
    "value": "one",
    "integration_id": 1,
    "claimed_by": 1,
    "createdAt": "2019-04-28T19:01:07.660Z",
    "updatedAt": "2019-04-28T19:12:03.643Z",
    "integrationId": 1,
    "integration": {
        "id": 1,
        "name": "test",
        "code": "test",
        "quota_period": "month",
        "quota_amount": 1,
        "description": "test",
        "createdAt": "2019-04-28T19:01:01.724Z",
        "updatedAt": "2019-04-28T19:01:01.724Z"
    }
}]
```

## It only affects Sequelize 5.x, not 4.x

I've tried to downgrade to sequelize and sequelize-cli 4.x, and it works as expected:

```bash
$ npm ls | grep sequelize
├─┬ sequelize@4.43.2
├─┬ sequelize-cli@4.1.1
```

and the server response (with the exact same example above, only thing changed was downgrading `sequelize` and `sequelize-cli` packages) looks like this:

```js
[{
    "id": 8,
    "value": "3",
    "integration_id": 2,
    "claimed_by": 1,
    "created_at": "2019-04-28T19:14:51.641Z",
    "updated_at": "2019-04-29T16:52:31.198Z",
    "integration": {
        "id": 2,
        "name": "Help",
        "code": "flixbus",
        "quota_period": "month",
        "quota_amount": 3,
        "description": "**I need somebody**",
        "created_at": "2019-04-28T19:14:15.164Z",
        "updated_at": "2019-04-28T19:14:15.164Z"
    }
}]
```

## Environment
Dialect: postgres, also can be that it affects all of them.
Dialect **library** version: pg 7.10.0
Database version: psql 10
Sequelize version: 5.7.6
Node Version: 10.2.1
OS: MacOS Mojave 10.14.4 (x64)
Tested with latest release: yes, with 5.7.6 (latest NPM release afaik)

**Comment by serge1peshcoff** [comment]
Also, this can help probably, here's one of my SQL log lines, seems like it queries the right fields but maps them to wrong values (`camelCase` ones instead of `snake_case` ones):

```
Executing (default): SELECT "id", "name", "code", "quota_period", "quota_amount", "description", "created_at" AS "createdAt", "updated_at" AS "updatedAt" FROM "integrations" AS "integration";
```

**Comment by aaronovz1** [comment]
I am seeing the same problem. My issue seems to be that only **association**/foreign key variables are not becoming snake_case (same as #10581).

**Comment by serge1peshcoff** [comment]
@aaronovz1 so the timestamps are returned in `snake_case` for you? Also, what version of Sequelize are you using? (It can be that it affects latest Sequelize versions only, just want to check it).

**Comment by aaronovz1** [comment]
@serge1peshcoff actually the timestamps are camelCase too. Looks like we have the exact same problem. I am running the Sequelize 5.7.6.

**Comment by serge1peshcoff** [comment]
@aaronovz1 then we are on the same page and I'm not only one facing this :D

Btw, I've tried to downgrade to `sequelize` and `sequelize-cli` 4.x, and it works as expected:

```bash
$ npm ls | grep sequelize
├─┬ sequelize@4.43.2
├─┬ sequelize-cli@4.1.1
```

and the server response (with the exact same example) looks like this:

```js
[{
    "id": 8,
    "value": "3",
    "integration_id": 2,
    "claimed_by": 1,
    "created_at": "2019-04-28T19:14:51.641Z",
    "updated_at": "2019-04-29T16:52:31.198Z",
    "integration": {
        "id": 2,
        "name": "Help",
        "code": "flixbus",
        "quota_period": "month",
        "quota_amount": 3,
        "description": "**I need somebody**",
        "created_at": "2019-04-28T19:14:15.164Z",
        "updated_at": "2019-04-28T19:14:15.164Z"
    }
}]
```

**Comment by aaronovz1** [comment]
I only recently upgraded to 5.x, so it is definitely a bug in V5 as you've just proven by downgrading. Hope this can be fixed ASAP.

**Comment by serge1peshcoff** [comment]
@aaronovz1 can you try reproducing it with other SQL dialects? I want to check if it affects only Postgres or others as well (I bet it's the latter but I can't check on my PC)

**Comment by aaronovz1** [comment]
Unfortunately I don't have the time or means to get other dialects setup at the moment.

**Comment by gdanov** [comment]
I have the same or similar issue with sequelize v5 and postgres. My schema is smaller so I'm seing this duplication only on xxxxx_id fields (and all of them are relations). Here are few examples:

`Executing (default): SELECT 
"users"."id", 
"users"."username", 
"users"."password_hash", 
"users"."account_id", 
"users"."account_id" AS "accountId", 
"account"."id" AS "account.id", 
"account"."name" AS "account.name", 
"account"."token_balance" AS "account.token_balance" 
FROM "users" AS "users" LEFT OUTER JOIN "account" AS "account" ON "users"."account_id" = "account"."id" WHERE "users"."account_id" = 1 LIMIT 1;`

`Executing (default): SELECT "id", "username", "password_hash", "account_id", "account_id" AS "accountId" FROM "users" AS "users" WHERE "users"."id" = 1;`

this one shows that the 'camel-ization' happens only on the root entity table (request_log) while the JOIN-ed table is not affected (see report_request.account_id)
`Executing (default): SELECT 
"request_log"."id", 
"request_log"."report_request_id", 
"request_log"."report_request_id" AS "reportRequestId", 
"request_log"."ts", 
"request_log"."user_id", 
"report_request"."id" AS "report_request.id", 
"report_request"."account_id" AS "report_request.account_id", 
"report_request"."was_billed" AS "report_request.was_billed", 
"report_request"."exec_ts" AS "report_request.exec_ts", 
"report_request"."ttl_ts" AS "report_request.ttl_ts", 
"report_request"."balance_snapshot" AS "report_request.balance_snapshot" 
FROM "request_log" AS "request_log" 
LEFT OUTER JOIN "report_request" AS "report_request" ON "request_log"."report_request_id" = "report_request"."id" 
WHERE "request_log"."user_id" = '1' ORDER BY "request_log"."ts" DESC LIMIT '10' OFFSET '0';`

**Comment by Strate** [comment]
Almost same issue here: I use underscored: true & timestamps: true, and I expect `created_at`, `updated_at` fields in snake_case, as it was on sequelize@4. But after upgrading to sequelize@5 this becomes blocker.

**Comment by serge1peshcoff** [comment]
@Strate same issue as I'm facing then, it affects timestamps as well as foreign key fields. My timestamps are also in `camelCase`

**Comment by serge1peshcoff** [comment]
I've recently understood why it's failing.
Each model has an `attributes` object, where the keys are the `camelCase` attributes names, and each of these objects has a `field` field which is a `snake_case` or `camelCase` (depending on if `underscored` is `true` or `false`).
In the `findAll()` method where the options object for SQL query is constructed, it uses `Object.keys(this.rawAttributes)` which will take the names in `camelCase` disregarding the `underscored` option, here: https://github.com/sequelize/sequelize/blob/30c5ca5e127474f585388a9f26aefc06024fe2ec/lib/model.js#L1732

If I change it to `Object.keys(this.rawAttributes).map(key => this.rawAttributes[key].field)` it should work. However it won't turn includes attributes to `snake_case`.

I'll most like make a PR on this either this or next week, if somebody will be faster then me I'd really appreciate it!

**Comment by Strate** [comment]
@serge1peshcoff hi, would make a PR to sequelize?

**Comment by serge1peshcoff** [comment]
@Strate I still haven't figure out how to do it properly, do you know how?

**Comment by Strate** [comment]
@serge1peshcoff yes, use this link: https://github.com/sequelize/sequelize/compare/master...serge1peshcoff:fix-not-respecting-underscored-in-findall

**Comment by serge1peshcoff** [comment]
@Strate yeah, for me it breaks half of the tests, that's why I didn't do a PR yet (and also because I really lack free time atm)

**Comment by konncradle** [comment]
Version 6.x seems to be affected as well.

**Comment by serge1peshcoff** [comment]
@sushantdhiman are there any plans or ETA to fix it?

**Comment by twistedrc1017** [comment]
i am also experiencing this on v5.10.1
![image](https://user-images.githubusercontent.com/31547504/61503375-8f46f180-a9a5-11e9-9ee6-a7076149231d.png)

it seems fine when i look at my db with pgAdmin though:
![image](https://user-images.githubusercontent.com/31547504/61503324-5d358f80-a9a5-11e9-9189-aae875f7748b.png)

**Comment by lefterisk** [comment]
+1

**Comment by frastlin** [comment]
There are many issues referencing this problem:

https://github.com/sequelize/sequelize/issues/11062
https://github.com/sequelize/sequelize/issues/11149
https://github.com/sequelize/sequelize/issues/10971
https://github.com/sequelize/sequelize/issues/11037
https://github.com/feathersjs-ecosystem/feathers-sequelize/issues/287

I'm not sure if this has anything to do with this bug:
https://github.com/sequelize/sequelize/issues/10977

Then I am having this problem as well.

**Comment by papb** [comment]
Closing in favor of #11225.

When #11225 gets closed, if anyone thinks this particular issue is still an issue, please ping me.

**Comment by tukno** [comment]
Hello, 
I was having the same problem as @twistedrc1017 with the timestamps `created_ at` and `updated_at`.  
The timestamps fields were both defined as snake case in the model:

![image](https://user-images.githubusercontent.com/14236708/65474166-d2f82480-de48-11e9-8a62-8ac49f594364.png)

And both undescored and undesrcoredAll were set to `true` in my database configuration file:

![image](https://user-images.githubusercontent.com/14236708/65474803-210e2780-de4b-11e9-8557-0d56963a854a.png)

But I was still getting cameCase timestamps in my responses:

![image](https://user-images.githubusercontent.com/14236708/65474097-8f051f80-de48-11e9-9c03-1a9068487285.png)

![image](https://user-images.githubusercontent.com/14236708/65474687-cbd21600-de4a-11e9-8288-0345801ce447.png)

To get my timestamps to snake case I set the following values to createdAt:'created_at' and the updatedAt:'updated_at' in my database config (as said in the docs [If you want sequelize to handle timestamps, but only want some of them, or want your timestamps to be called something else, you can override each column individually](https://sequelize.org/master/manual/models-definition.html#configuration) ):

![image](https://user-images.githubusercontent.com/14236708/65474317-544fb700-de49-11e9-81fd-b16b273acfe4.png)

 Then I got my timestamps in snake case:

![image](https://user-images.githubusercontent.com/14236708/65474418-b90b1180-de49-11e9-8c97-06c8799cc72b.png)

![image](https://user-images.githubusercontent.com/14236708/65474726-e6a48a80-de4a-11e9-8484-a8141102fb2d.png)

All other fields were already being converted from camelCase to snake case and coming back in response as snake case. The only fields that were coming in response as camelCase were the two timestamps mentioned, the configuration helped solve my issue, I hope it helps you too! :grimacing:

**Comment by jramiresbrito** [comment]
> Almost same issue here: I use underscored: true & timestamps: true, and I expect `created_at`, `updated_at` fields in snake_case, as it was on sequelize@4. But after upgrading to sequelize@5 fields becomes `createdAt` and `updatedAt`

Exactly the same to me. Here's the error:

`Executing (default): INSERT INTO "users" ("id","name","email","password_hash","created_at","updated_at") VALUES (DEFAULT,$1,$2,$3,$4,$5) RETURNING *;
(node:9418) UnhandledPromiseRejectionWarning: SequelizeDatabaseError: column "created_at" of relation "users" does not exist`

**Comment by sorushsaghari** [comment]
hey guys . it's still not working . 
i have __create_at__ in migration and used **_underscored: true_** in my models but it's not working

**Comment by DanTheAngell** [comment]
I've just upgraded from version 4 to 5.21.5.  It looks like `underscored: true` no longer honors foreign key conventions.  For example, I had to explicitly add `foreignKey` to make this work the way it used to:

```
Feature.belongsToMany(models.permission, {
	through: models.feature_permission,
	foreignKey: "feature_id"
});
```

**Comment by miladmeidanshahi** [comment]
option `underscored: true` affected on database column names, after find quarry still is camel case.

**Comment by kd-krishna** [comment]
im using version 5.21.4 and noticed that in case of many to many association , Im  my snake_case for my columns. when  trying to insert value in the joined table using add method , the internal query seems to be still looking for camelCase foreignkey column instead snake_case.  Any suggestions on how to get it working?

**Comment by miladmeidanshahi** [comment]
@papb any idea when this bug will be fixed?

**Comment by tr-fi** [comment]
5.21.5 same problem

**Comment by keephopealive** [comment]
Same problem here with sequelize version 5.21.6
Only seems to not apply 'underscore:true' in the association model.

**Comment by keephopealive** [comment]
> I've just upgraded from version 4 to 5.21.5. It looks like `underscored: true` no longer honors foreign key conventions. For example, I had to explicitly add `foreignKey` to make this work the way it used to:
> 
> ```
> Feature.belongsToMany(models.permission, {
> 	through: models.feature_permission,
> 	foreignKey: "feature_id"
> });
> ```
> 
> If I don't call it out it becomes `featureId`.

Great work around, works for now and when this feature is fixed, as long as we keep `underscore: true` in place, it should work either way.

**Comment by Aldiwildan77** [comment]
this error still appear on version 6.3.4
I've added underscore: true for each model but still doesn't work

**Comment by lucasnr** [comment]
I'm having the same issue. Any updates on it?

**Comment by datanotion** [comment]
Any idea why this was closed when it's still an issue?

**Comment by kaio-giovanni** [comment]
^6.3.5 same problem

**Comment by andreidmour** [comment]
6.3.5 confirmed bug

**Comment by GorgeousPuree** [comment]
It's just hilarious. Fortunately, only after 20 mins of struggling I've realized that's this was not my fault. 
^6.3.5 confirmed bug.

**Comment by tchestnut85** [comment]
I've also noticed this issue. Any updates?

**Comment by xtianlt** [comment]
timestamps: false

Please try this, but disable other functionality.

**Comment by viniciustrindade** [comment]
Same bug here at sequelize@6.3.5

**Comment by max107** [comment]
My grandchildren have already gone to school. But this error has still not fixed.

**Comment by DanTheAngell** [comment]
I think it's worth noting that based on the current documentation, this is not a bug at all but instead a fundamental change to how it works: 

https://sequelize.org/master/manual/naming-strategies.html

> Note that in both cases the fields are still camelCase in the JavaScript side; this option only changes how these fields are mapped to the database itself. The field option of every attribute is set to their snake_case version, but the attribute itself remains camelCase.

Which is great if I'm starting a new project from scratch. I think fully camelCased Javascript field names is a great convention; but it causes real problems when upgrading older Sequelize projects where we expect our Javascript attributes in snake_case. As mentioned above, this can be mitigated by explicitly setting any autogenerated fields like `createdAt: 'created_at'`.

But I wouldn't expect there to ever be a "fix" since it now matches up with their documentation.

**Comment by liuyib** [comment]
> I've just upgraded from version 4 to 5.21.5. It looks like `underscored: true` no longer honors foreign key conventions. For example, I had to explicitly add `foreignKey` to make this work the way it used to:
> 
> ```
> Feature.belongsToMany(models.permission, {
> 	through: models.feature_permission,
> 	foreignKey: "feature_id"
> });
> ```
> 
> If I don't call it out it becomes `featureId`.

Great, what you said is the root cause

**Comment by feeh27** [comment]
Thanks @dzvid your configuration worked for me:

![image](https://user-images.githubusercontent.com/13547485/111826790-e104f500-88c7-11eb-98e4-563692b45538.png)

**Comment by miladmeidanshahi** [comment]
As Sequelize said: https://sequelize.org/master/manual/naming-strategies.html
``
``

**Comment by hwookim** [comment]
Sadly, they don't think it's a bug...
I'd probably just use `createdAt: created_at`
I'm on `^6.6.2`

**Comment by vkbishu** [comment]
It's a bug and It should be fixed. As it's not a good idea to use 2 different cases for the field names.

**Comment by imaccallum** [comment]
Here's what it was for me: we have a join table that uses camel case keys. There was a belongsToMany relationship in the corresponding other models that did not include `foreignKey` and `otherKey`. Presumably if these are not specified it defaults to the camel case version and adds it to the join table or something weird like that

**Comment by liorsion** [comment]
This issue presents an even bigger problem when using TypeScript - we can't tell the compiler which field name is used because it changes..

**Comment by ephys** [comment]
I'm trying to understand this issue a bit better. Is the issue: 

1. That you want to use `created_at`, `user_id` on the JS side, but sequelize generates camelCase,
2. or that you want to use `createdAt`, `userId` on the JS side, but have `created_at`, `user_id` on the database side.

**Comment by miladmeidanshahi** [comment]
Hi @ephys absolutely number 1 :)

**Comment by ephys** [comment]
@miladmeidanshahi Thanks!

For clarity's sake, in the following text when I say:
- attribute, I mean the JavaScript name,
- column, I mean the Database name.

So by default Sequelize will generate camelCase attributes and snake_case column ([actually that second part is not the default yet](https://github.com/sequelize/sequelize/issues/13896)) names because we consider it to be the standard naming conventions of JS & SQL respectively.

The `underscored` option is only used to generate snake_case column names. As others have found out in this thread, it has no impact on attribute names.

For *timestamp attributes*, the solution is simple enough. You can configure it per-model using the 3 timestamp options:
```typescript
UserModel.init({ /* attributes */ }, {
  createdAt: 'created_at',
  updatedAt: 'updated_at',
  deletedAt: 'deleted_at',
})
``` 

Or configure them for every model by configuring the Sequelize instance itself:

```typescript
new Sequelize({
  define: {
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    deletedAt: 'deleted_at',
  }
});
```

For *foreign keys*, it's trickier. There is a way to configure it per-model but not system wide configuration option.

```typescript
UserModel.hasMany(Film, {
  as: 'author',
  foreignKey: {
    name: 'author_id', // the JavaScript attribute name
    field: 'author_id', // the column name
  },
})
```

To automatically generate a snake_case foreign key attribute names would require the return of an option like `underscoredAll` (but with a better name).

If a way to tell sequelize to generate attributes in snake_case is the desired solution, I'll keep this closed in favor of https://github.com/sequelize/sequelize/issues/11225 as it seems to be the relevant thread.

Let me know what you think :)

**Comment by hmarat** [comment]
@ephys Thanks, worked as expected :))

**Comment by yisibl** [comment]
@ephys After I override them globally, .update() will not work.

```js
new Sequelize({
  define: {
    createdAt: 'created_at',
    updatedAt: 'updated_at',
    deletedAt: 'deleted_at',
  }
});

// Model
Foo.prototype.change = async function() {
    const id = this.id
    return await this.update({
      updated_at: new Date(),
    }, {
      where: {
        id: id,
      },
    })
  }
```

I eventually found that the solution was to call `this.changed('updated_at', true)` once before `.update()`.

When I upgraded from v5 to v6, this became very buggy - when exactly did we need to call `.changed()`?

 At the end of the day, we just want to return a snake_case timestamp and want a unified API to solve this problem once and for all.

**Comment by yisibl** [comment]
https://github.com/sequelize/sequelize/issues/13006

**Comment by ephys** [comment]
Please open a separate bug report with a reproducible example as that is not related to this thread
